cudax/examples/cccl-rt/kernel_launch_patterns.cu

```diff
@@ -0,0 +1,254 @@
+    static_assert(decltype(config.dims)::static_extents(cudax::block, cudax::grid)[0] == 1);
+    static_assert(decltype(config.dims)::static_extents(cudax::block, cudax::grid)[1] == 1);
+  }
+
+  template <class Config>
+  __device__ void demo_dynamic_shared_memory(const Config& config)
+  {
+    // If the config contains a cudax::dynamic_shared_memory_option option, the cudax::dynamic_smem_ref function can be
+    // used to get a reference to the dynamic shared memory type. Keep in mind that the object is not constructed, so
+    // one of the threads must construct it before it is used.
+    auto& dyn_smem = cudax::dynamic_smem_ref(config);
+
+    // Construct the dynamic_smem_layout object in the shared memory by the first thread in the block.
+    if (config.dims.rank(cudax::thread, cudax::block) == 0)
+    {
+      new (&dyn_smem) dynamic_smem_layout{42};
+    }
+
+    // Wait until the write is finished.
+    __syncthreads();
+
+    // All threads should see the same value in the shared memory.
+    assert(dyn_smem.value == 42);
+  }
+};
+
+int main()
+try
+{
+  // Check we have at least one device.
+  if (cuda::devices.size() == 0)
+  {
+    std::fprintf(stderr, "No CUDA devices found\n");
+    return 1;
+  }
+
+  // We will use the first device.
+  cuda::device_ref device = cuda::devices[0];
+
+  // cudax::launch always requires a work submitter, so let's create a CUDA stream.
+  cuda::stream stream{device};
+
+  // Create a custom hierarchy to be used in cudax::launch. We will be launching a 1D grid of 1 block. The block will be
+  // a 2D grid of 2 threads in x and y axis.
+  //
+  // Note that the grid dimensions are passed as template parameters in this example. That means the value can be used
+  // in constexpr context inside the kernel. Block dimensions will be constructed at runtime as usually.
+  const auto hierarchy = cudax::make_hierarchy(cudax::grid_dims<1>(), cudax::block_dims(dim3{2, 2}));
+
+  // Launch an ordinary kernel. cudax::launch takes a stream as the first argument followed by the kernel configuration,
+  // kernel and kernel parameters.
+  cudax::launch(stream, cudax::make_config(hierarchy), kernel, "kernel");
+
+  // Launch a kernel functor. Here, we use cudax::distribute to create the kernel_config for us. This function creates
+  // a simple 1D grid of 1D blocks of a given size.
+  cudax::launch(stream, cudax::distribute<4>(4), kernel_functor{42}, name_buffer{"kernel functor"});
+
+  // Launch a kernel functor that takes a cudax::kernel_config. Note that the kernel config is passed automatically as
+  // the first argument by the cudax::launch function.
+  //
+  // The kernel functor requires dynamic memory, so we need to create the kernel configuration with dynamic shared
+  // memory option. The config remembers the type passed inside the option and makes it the return type of the
+  // cudax::dynamic_smem_ref(config) call inside the device code. See demo_dynamic_shared_memory for more information.
```

Comment:
suggestion: Lets fork off the `dynamic_shared_memory` to a separate example where we can dive into the nuance required here. To properply motivate this, we first need to explain why you can't just write code like this:

```
template <typename T>
__global__ void foo(T i){
    extern __shared__ T dyn_shmem[];
}
```